catalog: add support for relative targets and implicit types in Location entities

freben opened the pull request:
It was annoying me that it was so hard to do tests with the example data without rewriting ALL of the locations based on whether they were local or on github.

This is making a previously required field of a not-documented kind optional. Considering not honouring a period of notice for this one. Thoughts?

Fox32 commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -884,3 +885,58 @@ This kind is not yet defined, but is reserved [for future use](system-model.md).
+relative target paths (see below), and it will work out no matter if it's
+consumed locally on disk from a `file` location, or as uploaded on a VCS.
+
+### `spec.target` [optional]
```
Do we really need both a target and a targets variant?

Fox32 replied:
But as it was always there, it's a different topic 😉

freben replied:
Agreed that it seems excessive to have, only as a minor convenience to the file writer :) Maybe one day we deprecate it

Fox32 commented:
Documenting the Location entity? Great!
Adding relative targets 🤯 
This is so annoying, my `app-config.local.yaml` is full with locations to make this work... Thanks!

Rugvip commented on plugins/catalog-backend/src/ingestion/processors/LocationEntityProcessor.ts:
```diff
@@ -17,26 +17,39 @@
+  try {
+    if (base.type === 'file') {
+      if (target.startsWith('.')) {
+        return path.join(path.dirname(base.target), target);
```
Prolly wanna check that `base.target` doesn't end with a '/' or equivalent before grabbing the dirname right? I'm thinking that a base location of `/root/data/` with a target of `./catalog-info.yaml` would resolve to `/root/data/catalog-info.yaml`. Not really technically possible to have that happen right now but thinking it makes sense to handle correctly.

freben replied:
So to be clear, SHOULD that result in `/root/data/catalog-info.yaml` or not? I'm actually not sure :) I'm usually of the opinion that trailing slashes should have no effect

Rugvip replied:
Could throw an error if there's a trailing path separator for now, that way we get to decide later and it's not needed right now anyway :grin:

freben replied:
That sounds wise.

freben replied:
@Rugvip done!

freben replied:
(below)

Rugvip reviewed (commented):
Yuuuuuuuuuuuuuuuup, noice 😎 

Should do an extra round of thing around the security of relative file locations, but at a glance I'm thinking it's fine as we can't really escalate access with it.

freben commented on plugins/catalog-backend/src/ingestion/processors/LocationEntityProcessor.ts:
```diff
@@ -17,27 +17,52 @@
-          emit(result.location({ type: location.spec.type, target }, false));
-        }
+
+      const targets = new Array<string>();
```
it is what it is

Rugvip commented on packages/catalog-model/examples/acme-corp.yaml:
```diff
@@ -4,6 +4,5 @@ metadata:
-  type: github
   targets:
-    - https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/acme/org.yaml
+    - ./acme/org.yaml
```
Can't do this update and the other ones yet right? Would be breaking all existing usage in apps. I think giving the regular two weeks would be reasonable

Rugvip commented on .changeset/olive-parrots-retire.md:
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/catalog-model': minor
```
If the changes are they we start supporting relative paths and make the type optional, this could just be a patch

andrewthauer commented:
This seems very handy and I could definitely have used this already a few times 😁

freben commented:
@Rugvip That's a good point; even though it's just example data it's actively loaded by existing and new installations. I reverted the examples, put a README in place with a warning, and added a new `examples-relative` dir in there for those that are eager to use it before that.

I think I'm done now, and will add a do-not-merge PR for moving the new data into the old location.
